x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Sample.java

```diff
@@ -0,0 +1,164 @@
+import org.elasticsearch.xpack.esql.planner.ToAggregator;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+
+public class Sample extends AggregateFunction implements ToAggregator {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Sample", Sample::new);
+
+    @FunctionInfo(
+        returnType = {
+            "boolean",
+            "cartesian_point",
+            "cartesian_shape",
+            "date",
+            "date_nanos",
+            "double",
+            "geo_point",
+            "geo_shape",
+            "integer",
+            "ip",
+            "keyword",
+            "long",
+            "version" },
+        description = "Collects sample values for a field.",
+        type = FunctionType.AGGREGATE,
+        examples = @Example(file = "stats_sample", tag = "doc")
+    )
+    public Sample(
+        Source source,
+        @Param(
+            name = "field",
+            type = {
+                "boolean",
+                "cartesian_point",
+                "cartesian_shape",
+                "date",
+                "date_nanos",
+                "double",
+                "geo_point",
+                "geo_shape",
+                "integer",
+                "ip",
+                "keyword",
+                "long",
+                "text",
+                "version" },
+            description = "The field to collect sample values for."
+        ) Expression field,
+        @Param(name = "limit", type = { "integer" }, description = "The maximum number of values to collect.") Expression limit
+    ) {
+        this(source, field, Literal.TRUE, limit);
+    }
+
+    public Sample(Source source, Expression field, Expression filter, Expression limit) {
+        this(source, field, filter, limit, new Literal(Source.EMPTY, Randomness.get().nextLong(), DataType.LONG));
+    }
+
+    /**
+     * The query "FROM data | STATS s1=SAMPLE(x,N), s2=SAMPLE(x,N)" should give two different
+     * samples of size N. The uuid is used to ensure that the optimizer does not optimize both
+     * expressions to one, resulting in identical samples.
+     */
+    public Sample(Source source, Expression field, Expression filter, Expression limit, Expression uuid) {
+        super(source, field, filter, List.of(limit, uuid));
+    }
+
+    private Sample(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        int limit = (int) limitField().fold(FoldContext.small());
+        if (limit <= 0) {
+            return new TypeResolution(format(null, "Limit must be greater than 0 in [{}], found [{}]", sourceText(), limit));
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
```

Comment:
There are some missing validations here:
- We should have here type validations for both parameters. Integer for the limit, and nearly any type but ulong and counters for the field
  - You can follow the `Max` example for the field: https://github.com/elastic/elasticsearch/blob/47edd0adb858e738b177c8b7e8564989415564ab/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Max.java#L125-L133
- Before folding the limit, you should check if it `isFoldable(limitField(), sourceText(), SECOND)`

I wonder why no test showed an error here though. At least for unsigned long (Unless this function can actually work with them; in which case it should be included in the function&field annotation

Reply:
We're not covering them much in functions yet, so it's fine as it's now, thanks!